### PR TITLE
Adding TraceContext as parameter

### PR DESCRIPTION
**Pull request by yojagad**
Part of fix for https://github.com/Azure/azure-functions-host/issues/3747

### CODE REVIEW

File: src/Public/TraceContext.cs [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+namespace Microsoft.Azure.Functions.PowerShellWorker
+{
+    /// <summary>
+    /// Custom TraceContext constructed from the TraceContext members in gRPC.
+    /// </summary>
+    public class TraceContext
+    {
+        /// <summary>
+        /// Constructor for TraceContext.
```
Comment (AnatoliB) [inline src/Public/TraceContext.cs]: Consider removing (auto-generated?) comments like this, as they don't add any value.

File: src/Public/TraceContext.cs [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        {
+            Traceparent = traceParent;
+            Tracestate = traceState;
+            Attributes = GetCaseInsensitiveAttributes(attributes);
+        }
+
+        /// <summary>
+        /// Gets the traceparent.
+        /// </summary>
+        public string Traceparent { get; }
```
Comment (AnatoliB) [inline src/Public/TraceContext.cs]: Casing: TraceParent, TraceState

File: src/Public/TraceContext.cs [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        /// </summary>
+        public TraceContext(string traceParent, string traceState, MapField<string, string> attributes)
+        {
+            Traceparent = traceParent;
+            Tracestate = traceState;
+            Attributes = GetCaseInsensitiveAttributes(attributes);
+        }
+
+        /// <summary>
+        /// Gets the traceparent.
```
Comment (AnatoliB) [inline src/Public/TraceContext.cs]: This is what the code says, directly and literally. Please remove the comment.

File: src/FunctionInfo.cs [hunk 4]
```diff
@@ -24,9 +24,12 @@ internal class AzFunctionInfo
         private const string ActivityTrigger = "activityTrigger";
 
         internal const string TriggerMetadata = "TriggerMetadata";
+        internal const string TraceContext = "TraceContext";
```
Comment (AnatoliB) [inline src/FunctionInfo.cs]: Please also update the design doc (https://github.com/Azure/azure-functions-powershell-worker/blob/dev/docs/designs/PowerShell-AzF-Overall-Design.md) as well: just a short note describing this parameter, perhaps with a usage sample.

File: src/Public/TraceContext.cs (hunk 1 above)
Comment (yojagad) [reply src/Public/TraceContext.cs]: I was seeing errors if these comments were not added.

**Comment by AnatoliB** [comment]
Is this PR missing the .proto file changes?

**Comment by AnatoliB** [comment]
The instructions on updating .proto are here: https://github.com/Azure/azure-functions-powershell-worker/blob/dev/protobuf/README.md

File: src/Public/TraceContext.cs (hunk 1 above)
Comment (AnatoliB) [reply src/Public/TraceContext.cs]: Oh, I see. One way to suppress these errors is to declare this class `internal`.

File: src/Public/TraceContext.cs (hunk 2 above)
Comment (AnatoliB) [reply src/Public/TraceContext.cs]: Apparently, there is a decision on these names that I was not aware of (https://github.com/Azure/azure-functions-host/pull/4917#discussion_r324471617), so please ignore my comment here.

File: src/FunctionInfo.cs (hunk 4 above)
Comment (yojagad) [reply src/FunctionInfo.cs]: Updated InvocationRequest section with some text and snippet of how to access TraceContext. Please let me know if anything more needs to be added there.

File: docs/designs/PowerShell-AzF-Overall-Design.md [hunk 5]
```diff
@@ -67,6 +67,19 @@ The host process sends an _InvocationRequest_ every time a function is triggered
 
 The real invocation happens mainly in the PowerShell Manager. It transforms the input binding data to the appropriate types that are easy to work with in PowerShell, and then invokes the function script with the transformed value as arguments. Once the invocation finishes, it collects the results for the output bindings. The output results will then be converted to the pre-defined gRPC data type, and sent to the Host within an _InvocationResponse_.
 
+#### TraceContext
+
+As part of the _InvocationRequest_, the host also sends TraceContext which contains the Traceparent (the current Activity id), Tracestate and Attributes (The current map of tags from Activity). Users can access this by specifying it as parameter as shown in the example below.
```
Comment (AnatoliB) [inline docs/designs/PowerShell-AzF-Overall-Design.md]: Casing: "**t**he current map"

File: docs/designs/PowerShell-AzF-Overall-Design.md [hunk 6]
```diff
@@ -67,6 +67,19 @@ The host process sends an _InvocationRequest_ every time a function is triggered
...
+
+```powershell
+
+param($req, $TraceContext)
+
+$activityId = $TraceContext.Traceparent
```
Comment (AnatoliB) [inline docs/designs/PowerShell-AzF-Overall-Design.md]: Is there any specific tracing facility that the customer is supposed to be using with this data, or this is entirely up to the customer? I'm thinking about an code sample that would demonstrate the actual usage, not just retrieving properties into variables, e.g. `Write-Verbose "[$($TraceContext.Traceparent)] My message"`.

File: src/RequestProcessor.cs [hunk 7]
```diff
@@ -396,7 +397,12 @@ private Hashtable InvokeSingleActivityFunction(PowerShellManager psManager, AzFu
                 }
             }
 
-            return psManager.InvokeFunction(functionInfo, triggerMetadata, invocationRequest.InputData);
+            if(functionInfo.HasTraceContextParam)
```
Comment (AnatoliB) [inline src/RequestProcessor.cs]: Please insert a space between `if` and `(`.

File: docs/designs/PowerShell-AzF-Overall-Design.md (hunk 6 above)
Comment (yojagad) [reply docs/designs/PowerShell-AzF-Overall-Design.md]: Update text and pointed to https://www.w3.org/TR/trace-context/ as discussed

**Pull request merged**